### PR TITLE
Improve `isassigned` implementation

### PR DESCRIPTION
**Pull request by Tokazama**
Unless `isassigned` is called on `Array` with `Int`s, it uses a try catch, which makes `isassigned` pretty slow (#44720). This PR provides changes the default implementation of `isassigned` to coerce the indices provided to `Int`s and converts to linear or cartesian indices, depending on the arrays `IndexStyle`. This also overloads `isassigned` for many of the array types defined in Base.

### CODE REVIEW

**Review by vtjnash — COMMENTED** [review]
SGTM

**Comment by Tokazama** [comment]
I'm not sure if the errors I'm seeing are related to this PR (passes locally on my computer), but I'd be happy to make changes as needed

File: base/range.jl [hunk 1]
```diff
@@ -901,6 +901,8 @@ end
 
 ## indexing
 
+isassigned(r::AbstractRange, i::Int) = firstindex(r) <= i <= lastindex(r)
```
Comment (vtjnash) [inline base/range.jl:904]: Is this true for StepRange? I think it might only be true for Unit ranges

Comment (Tokazama) [reply base/range.jl:904]: I'm not aware of an instance where it could produce an undefined variable

Comment (vtjnash) [reply base/range.jl:904]: oh, right, I was thinking of values, not index

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 2]
```diff
@@ -604,6 +617,19 @@ function diag(M::Tridiagonal{T}, n::Integer=0) where T
     end
 end
 
+@inline function Base.isassigned(A::Tridiagonal, i::Int, j::Int)
+    @boundscheck checkbounds(A, i, j)
```
Comment (longemen3000) [inline stdlib/LinearAlgebra/src/tridiag.jl:621]: is this ok? in all other `isassigned` implementations, this returns false if the bound check fails, but this throws an error?

Comment (Tokazama) [reply stdlib/LinearAlgebra/src/tridiag.jl:621]: You're right. That should be fixed

**Pull request merged**